----------------------
### Obj : to create a simple ML model using linear regression
- on adversting dataset
------------------

In [1]:
import pandas as pd
import numpy as np
# import plotting libraries
import matplotlib.pyplot as plt

import seaborn as sns
sns.set(style="white", color_codes=True)
sns.set(font_scale=1.5)

# import the ML algorithm - step 1
from sklearn.linear_model import LinearRegression

#### load the Adv data

In [2]:
location = "https://github.com/gridflowai/gridflowAI-datasets-icons/raw/master/AI-DATASETS/01-MISC/Advertising.csv"

In [3]:
# load the training data from glass data set
df_training = pd.read_csv(location)

In [4]:
df_training

Unnamed: 0     TV  radio  newspaper  sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

In [5]:
df_training.drop(['Unnamed: 0'], axis=1, inplace=True)

In [6]:
df_training.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [64]:
df_training.shape

(200, 4)

#### Set the columns for X and y

In [7]:
# create a Python list of feature names
feature_cols = ['TV', 'radio', 'newspaper']
#feature_cols = ['TV']

In [8]:
# prepare dataset split
X = df_training[feature_cols]
y = df_training['sales']

#### Split the data in train and test parts

In [9]:
# now we will split the X and y into train test split
from sklearn.model_selection import train_test_split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.20, random_state=11, shuffle=True)

In [11]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((160, 3), (40, 3), (160,), (40,))

#### model creation

In [12]:
# instantiate - Step 2
linreg = LinearRegression()

In [13]:
# Fit , provie the data (X, y) - Training the model - Step 3
linreg.fit(X_train, y_train)

LinearRegression()

In [72]:
linreg.coef_, linreg.intercept_

(array([ 0.04342121,  0.19662236, -0.00726921]), 3.3052832530865857)

In [14]:
# Step 4

X1 = 150
X2 = 270
X3 = 20

# how much sales will happen
y_pred = 0.04576465*X1 + 0.18853002*X2 + -0.00103749*X3 + 2.9388893694594067
y_pred

60.685942469459405

In [15]:
linreg.predict([[X1, X2, X3]])
#linreg.predict([[X1]])

D:\anaconda\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([62.76111744])

In [75]:
# predict for the training (X_train) samples
#y_pred_training = linreg.predict(X_train)

In [76]:
# plt.figure(figsize=(6,4))
# plt.scatter(X_train, y_train)
# plt.scatter(X_train, y_pred_training, c='r')

# plt.xlabel('TV')
# plt.ylabel('sales');

#### Evaluate the model

In [16]:
y_pred = linreg.predict(X_test)

In [17]:
y_pred[:5]

array([13.53545074, 18.76822414, 11.92038864,  9.58208872,  6.64418613])

In [79]:
y_test[:5]

174    11.5
33     17.4
173    11.7
186    10.3
22      5.6
Name: sales, dtype: float64

In [18]:
from sklearn import metrics

In [19]:
print("the model MAE is = {:8.2f}".format(metrics.mean_absolute_error(y_test, y_pred)))

the model MAE is =     1.61


In [83]:
df_training.describe()

TV       radio   newspaper       sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   14.022500
std     85.854236   14.846809   21.778621    5.217457
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   10.375000
50%    149.750000   22.900000   25.750000   12.900000
75%    218.825000   36.525000   45.100000   17.400000
max    296.400000   49.600000  114.000000   27.000000

In [20]:
print("the model MSE is = {:8.2f}".format(metrics.mean_squared_error(y_test, y_pred)))

the model MSE is =     5.18


In [87]:
print("the model MSE is = {:8.2f}".format(np.sqrt(metrics.mean_squared_error(y_test, y_pred))))

the model MSE is =     2.28


In [21]:
print('MAPE                    : ', np.mean(np.abs((y_test - y_pred) / y_test)) * 100)

MAPE                    :  26.969730492520476


Meaning of MAPE : typical error % is 26.96

In [22]:
print('MPE                     : ', np.mean((y_test - y_pred) / y_test) * 100)

MPE                     :  -16.770327089937446


In [23]:
print('r2 value                : ', metrics.r2_score(y_test, y_pred))

r2 value                :  0.8449174816183896


meaning of 0.84 is - Linear reg model removed 84% of the error